### PR TITLE
refactored intersections.metta  func

### PR DESCRIPTION
**Pull request by Nardos-serkalem**

### CODE REVIEW

File: utilities/tree-helpers.metta [hunk 1]
```diff
@@ -362,3 +362,25 @@
...
+            ( 
+                ($firstChildGuardSet (getGuardSet $firstChild)) 
+                ($remainingIntersections (intersections $restChildren)) 
+                ($result (setIntersection treeIsMember $firstChildGuardSet $remainingIntersections)) 
+            ) 
+        $result 
+    ) 
+) 
+)
+;; testcase
```
Comment (Yagth) [inline utilities/tree-helpers.metta:383]: Remove this test cases from this file. Intersections' test file is found in the `enf-reducer/tests/helper-functions` add the following test cases and also some more if you can.
```scheme
(intersections Nil) 
(intersections (Cons (TreeNode (Value Nil False AND) (Cons (TreeNode (Value A True LITERAL) Nil Nil) Nil) Nil) (Cons (TreeNode (Value Nil False AND) (Cons (TreeNode (Value A True LITERAL) Nil Nil) Nil) Nil) Nil)))
```

Writing test cases like the one above isn't the only thing you need to do. Take a look in the file and make sure you use `assertEqual` to write your test cases.

**Review by Yagth — CHANGES_REQUESTED** [review]
I found the issue. You didn't remove the previous `intersections` function after you implemented the new one. Make sure to do that, add your test cases in the appropriate file I told you and  then test it locally. If everything is well, update your PR :)

File: utilities/tree-helpers.metta [hunk 2]
```diff
@@ -137,24 +137,22 @@
...
-    (
-      (0 Nil)
-      (1 (getGuardSet (head $children)))
-      ($else 
-        (let* 
-          (
-            ($firstChildGuardSet (getGuardSet (head $children)))
-            ($remainingIntersections (intersections (tail $children)))
+(= (intersections Nil) Nil)
+
```
Comment (Yagth) [inline utilities/tree-helpers.metta:143]: Remove Unnecessary spaces between a function definition to be consistent with our code base.

File: utilities/tree-helpers.metta [hunk 3]
```diff
@@ -137,24 +137,22 @@
 
 ;; Given list of trees, intersections will compute the intersection of between their guard sets.
 ;; It has similar behaviour like setIntersection but for more than two trees.
+
```
Comment (Yagth) [inline utilities/tree-helpers.metta:140]: Remove this line too

File: utilities/tree-helpers.metta [hunk 4]
```diff
@@ -137,24 +137,22 @@
...
-      (1 (getGuardSet (head $children)))
-      ($else 
-        (let* 
-          (
-            ($firstChildGuardSet (getGuardSet (head $children)))
-            ($remainingIntersections (intersections (tail $children)))
+(= (intersections Nil) Nil)
+
+(= (intersections (Cons $firstChild Nil)) (getGuardSet $firstChild))
+
```
Comment (Yagth) [inline utilities/tree-helpers.metta:145]: Same here

File: utilities/tests/tree-helpers-test.metta [hunk 5]
```diff
@@ -0,0 +1,8 @@
+! (register-module! ../../../metta-moses-reduction)
+! (import! &self metta-moses-reduction:utilities:tree-helpers)
+
+!(assertEqual(intersections Nil) Nil )
```
Comment (Yagth) [inline utilities/tests/tree-helpers-test.metta:4]: Add a space between `assertEqual` and `(intersections ... )`

File: utilities/tree-helpers.metta [hunk 6]
```diff
@@ -137,24 +137,22 @@
...
+    (let*
+        (
+            ($firstChildGuardSet (getGuardSet $firstChild))
+            ($remainingIntersections (intersections (Cons $secondChild $restChildren)))
             ($result (setIntersection treeIsMember $firstChildGuardSet $remainingIntersections))
-          )
-          $result
         )
-      )
+
```
Comment (Yagth) [inline utilities/tree-helpers.metta:153]: Same here. Remove this line

File: utilities/tree-helpers.metta (hunk 4 above)
Comment (Yagth) [reply utilities/tree-helpers.metta:145]: Remove this line

**Review by Yagth — CHANGES_REQUESTED** [review]

**Review by Nardos-serkalem — COMMENTED** [review]
I have  implemented the requested changes and updated the code accordingly.

**Review by Yagth — APPROVED** [review]

**Pull request merged**